## 3.3.motif分析

我们根据结构变化区域进行motif分析，主要分为三个部分：

* de novo sequence motif discovery
* known structure motif
* de novo structure motif

### 1) 数据说明

我们以liulab的第四批Shape-seq数据为例进行说明。具体内容可参考文件/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new

08.sequence文件夹为de novo sequence motif discovery的目录；01.fasta，05.cmscan，06.motif_alignment_info为 known structure motif的目录；07.BEAM为de novo structure motif的目录

### 2) de novo sequence motif discovery

我们将结构变化区域分为上调(gini index of UV+>gini index of UV-)和下调(gini index of UV+<gini index of UV-)的两个部分进行motif富集。首先汇总结构变化区域的序列信息，然后利用MEME进行de novo sequence motif discovery。

#### 2a) 汇总结构变化区域的序列信息

这里我们将上调结构变化区域和下调结构变化区域分别汇总到一个文件，要求每个结构变化区域序列长度相同，我们取50nt固定长度，超过50nt的结构变化区域进行截断。

脚本参考/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/08.sequence/change_region_fasta.py

```python
import pandas as pd
path = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/08.sequence/UVR/'
shape_data_coil = pd.read_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/result/uvr8/summary_result_merge_50_1_new.csv',sep='\t')
shape_data_coil['up_down']= shape_data_coil['delta'].map(lambda x: 'up' if x>0 else 'down')
shape_data_coil['gene']= shape_data_coil['transcript_id'].map(lambda x:x.split('.')[0])
data_change =shape_data_coil.loc[(shape_data_coil['hit_f']>2)&(shape_data_coil['hit_z']>2),:]

data_change_up = data_change.loc[data_change['up_down']=='up',:]
data_change_up.index=range(len(data_change_up))
data_change_down = data_change.loc[data_change['up_down']=='down',:]
data_change_down.index=range(len(data_change_down))

f = open(path + 'merge_up.fasta', 'w')
for i in range(len(data_change_up)):
    transcript = data_change_up.loc[i, 'transcript_id']
    location = data_change_up.loc[i, 'location_start']
    nucleotide = data_change_up.loc[i, 'nucleotide']
    nucleotide = nucleotide.replace(',', '')[:50]
    up_down = data_change_up.loc[i, 'up_down']
    start = data_change_up.loc[i, 'start']
    f.write('>' + transcript + '_' + location + '_' + str(int(start)) +'_'+ up_down+ '\n')
    f.write(nucleotide + '\n')

f.close()

f = open(path + 'merge_down.fasta', 'w')
for i in range(len(data_change_down)):
    transcript = data_change_down.loc[i, 'transcript_id']
    location = data_change_down.loc[i, 'location_start']
    nucleotide = data_change_down.loc[i, 'nucleotide']
    nucleotide = nucleotide.replace(',', '')[:50]
    up_down = data_change_down.loc[i, 'up_down']
    start = data_change_down.loc[i, 'start']
    f.write('>' + transcript + '_' + location + '_' + str(int(start)) +'_'+ up_down+ '\n')
    f.write(nucleotide + '\n')

f.close()


path = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/08.sequence/WT/'
shape_data_coil = pd.read_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/result/col/summary_result_merge_50_1_new.csv',sep='\t')
shape_data_coil['up_down']= shape_data_coil['delta'].map(lambda x: 'up' if x>0 else 'down')
shape_data_coil['gene']= shape_data_coil['transcript_id'].map(lambda x:x.split('.')[0])
data_change =shape_data_coil.loc[(shape_data_coil['hit_f']>2)&(shape_data_coil['hit_z']>2),:]

data_change_up = data_change.loc[data_change['up_down']=='up',:]
data_change_up.index=range(len(data_change_up))
data_change_down = data_change.loc[data_change['up_down']=='down',:]
data_change_down.index=range(len(data_change_down))

f = open(path + 'merge_up.fasta', 'w')
for i in range(len(data_change_up)):
    transcript = data_change_up.loc[i, 'transcript_id']
    location = data_change_up.loc[i, 'location_start']
    nucleotide = data_change_up.loc[i, 'nucleotide']
    nucleotide = nucleotide.replace(',', '')[:50]
    up_down = data_change_up.loc[i, 'up_down']
    start = data_change_up.loc[i, 'start']
    f.write('>' + transcript + '_' + location + '_' + str(int(start)) +'_'+ up_down+ '\n')
    f.write(nucleotide + '\n')

f.close()

f = open(path + 'merge_down.fasta', 'w')
for i in range(len(data_change_down)):
    transcript = data_change_down.loc[i, 'transcript_id']
    location = data_change_down.loc[i, 'location_start']
    nucleotide = data_change_down.loc[i, 'nucleotide']
    nucleotide = nucleotide.replace(',', '')[:50]
    up_down = data_change_down.loc[i, 'up_down']
    start = data_change_down.loc[i, 'start']
    f.write('>' + transcript + '_' + location + '_' + str(int(start)) +'_'+ up_down+ '\n')
    f.write(nucleotide + '\n')

f.close()
```

输出结果WT样本的merge_down.fasta、merge_up.fasta和UVR8突变体的merge_down.fasta、merge_up.fasta。

#### 2b) MEME

我们利用MEME在线版进行motif分析，地址为http://meme-suite.org/tools/meme。

![5.png](/png/5.png)

我们上传merge_down.fasta、merge_up.fasta后提交运算。

![6.png](/png/6.png)

在得到的结果中我们根据Evalue进行筛选，Sites为找到的motif在几个结构改变区域中出现过，由此可以计算出该motif的覆盖率。我们选择覆盖率较高的motif，这里取大于4%。


### 3) known structure motif

我们将结构变化区域序列与Rfam中已知的motif进行对比，找到结构变化区域中包含的已知的结构motif。

#### 3a) 获取结构改变区域序列

我们将同一转录本结构改变区域的序列放在一个文件中，不需要对序列长度进行处理。

参考脚本/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/01.fasta/change_region_fasta.py

```python
import pandas as pd
path = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/01.fasta/UVR/'
data = pd.read_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/col_nouv/final.modified_unmodified', sep='\t')
shape_data_coil = pd.read_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/result/uvr8/summary_result_merge_50_1.csv',sep='\t')
shape_data_coil['up_down']= shape_data_coil['delta'].map(lambda x: 'up' if x>0 else 'down')
hit_level_coil_uv_f = pd.read_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/uvr8_nouv/cutoff.hit.group',sep='\t')
hit_level_coil_uv_f = hit_level_coil_uv_f.reset_index()
hit_level_coil_uv_f.columns =['group','transcript_id','modified.median','unmodified.median','modified.sum','unmodified.sum','hit_f']
hit_level_coil_uv_z = pd.read_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/uvr8_uv/cutoff.hit.group',sep='\t')
hit_level_coil_uv_z = hit_level_coil_uv_z.reset_index()
hit_level_coil_uv_z.columns =['group','transcript_id','modified.median','unmodified.median','modified.sum','unmodified.sum','hit_z']
shape_data_coil =pd.merge(pd.merge(shape_data_coil,hit_level_coil_uv_f[['transcript_id','hit_f']],on='transcript_id',how='left'),hit_level_coil_uv_z[['transcript_id','hit_z']],on='transcript_id',how='left')
shape_data_coil['gene']= shape_data_coil['transcript_id'].map(lambda x:x.split('.')[0])
data_change =shape_data_coil.loc[(shape_data_coil['hit_f']>2)&(shape_data_coil['hit_z']>2),:]
UVR_only_sample = pd.read_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/result/uvr8/Uvr_shape_only.txt',sep='\t',header=None)
data_change = data_change.loc[data_change.gene.isin(list(UVR_only_sample[0])),:]
data_change.index=range(len(data_change))

for i in range(len(data_change)):
    if i%1000==0:
        print(i)
    trasnscript = data_change.loc[i,'transcript_id']
    Nucleotide = list(list(data.loc[data['transcript_id']==trasnscript,'Nucleotide'])[0])
    start = int(data_change.loc[i,'start'])
    end = int(data_change.loc[i, 'end'])
    data_change.loc[i,'nucleotide']=''.join(Nucleotide[start:end+1])


data_5UTR=data_change
trasnscript_list=list(set(data_5UTR['transcript_id']))
gene = [i.split('.')[0] for i in trasnscript_list]

for i in range(len(trasnscript_list)):
    trasnscript=trasnscript_list[i]
    f = open(path+trasnscript+'.fasta','w')
    data_5UTR_ = data_5UTR.loc[data_5UTR['transcript_id']==trasnscript,:]
    data_5UTR_.index=range(len(data_5UTR_))
    for j in range(len(data_5UTR_)):
        start = int(data_5UTR_.loc[j, 'start'])
        end = int(data_5UTR_.loc[j, 'end'])
        location = data_5UTR_.loc[j, 'location_start']
        up_down =data_5UTR_.loc[j, 'up_down']
        seq = data_5UTR_.loc[j, 'nucleotide']
        f.write('>'+trasnscript+'_'+location+'_'+up_down+'_'+str(start)+'_'+str(end)+'\n')
        f.write(seq+'\n')
    f.close()


path = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/01.fasta/WT/'
data = pd.read_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/col_nouv/final.modified_unmodified', sep='\t')
shape_data_coil = pd.read_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/result/col/summary_result_merge_50_1.csv',sep='\t')
shape_data_coil['up_down']= shape_data_coil['delta'].map(lambda x: 'up' if x>0 else 'down')
hit_level_coil_uv_f = pd.read_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/col_nouv/cutoff.hit.group',sep='\t')
hit_level_coil_uv_f = hit_level_coil_uv_f.reset_index()
hit_level_coil_uv_f.columns =['group','transcript_id','modified.median','unmodified.median','modified.sum','unmodified.sum','hit_f']
hit_level_coil_uv_z = pd.read_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/col_uv/cutoff.hit.group',sep='\t')
hit_level_coil_uv_z = hit_level_coil_uv_z.reset_index()
hit_level_coil_uv_z.columns =['group','transcript_id','modified.median','unmodified.median','modified.sum','unmodified.sum','hit_z']
shape_data_coil =pd.merge(pd.merge(shape_data_coil,hit_level_coil_uv_f[['transcript_id','hit_f']],on='transcript_id',how='left'),hit_level_coil_uv_z[['transcript_id','hit_z']],on='transcript_id',how='left')
shape_data_coil['gene']= shape_data_coil['transcript_id'].map(lambda x:x.split('.')[0])
data_change =shape_data_coil.loc[(shape_data_coil['hit_f']>2)&(shape_data_coil['hit_z']>2),:]
UVR_only_sample = pd.read_csv('/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/result/col/WT_shape_only.txt',sep='\t',header=None)
data_change = data_change.loc[data_change.gene.isin(list(UVR_only_sample[0])),:]
data_change.index=range(len(data_change))

for i in range(len(data_change)):
    if i%1000==0:
        print(i)
    trasnscript = data_change.loc[i,'transcript_id']
    Nucleotide = list(list(data.loc[data['transcript_id'] == trasnscript, 'Nucleotide'])[0])
    start = int(data_change.loc[i,'start'])
    end = int(data_change.loc[i, 'end'])
    data_change.loc[i,'nucleotide']=''.join(Nucleotide[start:end+1])


data_5UTR=data_change
trasnscript_list=list(set(data_5UTR['transcript_id']))
gene = [i.split('.')[0] for i in trasnscript_list]

for i in range(len(trasnscript_list)):
    trasnscript=trasnscript_list[i]
    f = open(path+trasnscript+'.fasta','w')
    data_5UTR_ = data_5UTR.loc[data_5UTR['transcript_id']==trasnscript,:]
    data_5UTR_.index=range(len(data_5UTR_))
    for j in range(len(data_5UTR_)):
        start = int(data_5UTR_.loc[j, 'start'])
        end = int(data_5UTR_.loc[j, 'end'])
        location = data_5UTR_.loc[j, 'location_start']
        up_down =data_5UTR_.loc[j, 'up_down']
        seq = data_5UTR_.loc[j, 'nucleotide']
        f.write('>'+trasnscript+'_'+location+'_'+up_down+'_'+str(start)+'_'+str(end)+'\n')
        f.write(seq+'\n')
    f.close()
```

输出目录为/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/01.fasta/WT和/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/01.fasta/UVR，文件夹4中每个转录本为一个文件。

#### 3b) cmscan软件对比Rfam

每个转录本都需进行Rfam，单个转录本进行对比的脚本如下：

```sh
echo AT3G16830.1 start `date`
/BioII/lulab_b/chenyinghui/software/infernal/infernal-1.1.2/src/cmscan 
/BioII/lulab_b/chenyinghui/database/Rfam/14.1/Rfam.cm 
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/01.fasta/WT/AT3G16830.1.fasta > 
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/05.cmscan/WT/AT3G16830.1.cmscan.txt
echo AT3G16830.1 end `date`
```
为了方便，我们写脚本批量生成所有转录本的比对脚本。

```sh
#!/bin/bash
source /BioII/lulab_b/containers/singularity/wrappers/bashrc
mkdir /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/05.cmscan/WT
python3 /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/05.cmscan/cmscan.py  \
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/01.fasta/WT  \
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/05.cmscan/WT   \
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/05.cmscan/WT.cmscan.sh

mkdir /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/05.cmscan/UVR
python3 /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/05.cmscan/cmscan.py  \
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/01.fasta/UVR  \
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/05.cmscan/UVR  \
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/05.cmscan/UVR.cmscan.sh
```

WT.cmscan.sh和UVR.cmscan.sh为所有转录本比对脚本，分别运行两个脚本即可完成与Rfam的比对，得到输出文件夹/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/05.cmscan/WT和/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/05.cmscan/UVR。

#### 3c) 汇总结果

过滤Evalue<1的motif，并进行汇总。

```sh
#!/bin/sh
# select hits whose E-value < 1

source /BioII/lulab_b/containers/singularity/wrappers/bashrc

mkdir -p /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/06.motif_alignment_info/WT

python /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/06.motif_alignment_info/cmscan_result_summary.sig.py \
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/05.cmscan/WT  \
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/06.motif_alignment_info/WT  \
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/06.motif_alignment_info/WT.motif.stat.xls

mkdir -p /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/06.motif_alignment_info/UVR
python /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/06.motif_alignment_info/cmscan_result_summary.sig.py \
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/05.cmscan/UVR  \
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/06.motif_alignment_info/UVR \
/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/06.motif_alignment_info/UVR.motif.stat.xls
```

输出文件/WT.motif.stat.xls和/UVR.motif.stat.xls包括Evalue>1的motif及其所在的转录本。

### 4) de novo structure motif

我们利用**BEAR软件**找到de novo structure motif，将结构变化区域分为上调(gini index of UV+>gini index of UV-)和下调(gini index of UV+<gini index of UV-)的两个部分进行motif富集。

我们的主要步骤分为：：

* 获取上调结构区域和下调结构区域对应的序列、UV-下reactivity、UV+下reactivity
* 基于**RNAstructure**，根据每个结构变化区域的序列和reactivity进行RNA二级结构预测，得到结构的DB文件。
* 将所有上调结构变化区域的序列和结构DB汇总到一个文件，将所有下调结构变化区域的序列和结构DB汇总到一个文件
* 基于BEAM进行motif富集分析
* 对找的的结构motif进行过滤和汇总

#### 4a) 结构区域对应的序列和reactivity

脚本参考/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/1.get_fasta_shape.py

```python
import pandas as pd

col_uv_f = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/col_nouv/'
col_uv_z = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/col_uv/'
uvr_uv_f = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/uvr8_nouv/'
uvr_uv_z = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/uvr8_uv/'

data_1 = pd.read_csv(col_uv_z + '/final.modified_unmodified_new', sep='\t', header=None)
data_1.columns = ['transcript_id', 'Nucleotide', 'Modified_mutations', 'Modified_effective_depth',
                     'Untreated_mutations', 'Untreated_effective_depth', 'R']
data_1 = data_1[['transcript_id', 'Nucleotide', 'Modified_mutations', 'Modified_effective_depth',
                       'Untreated_mutations', 'Untreated_effective_depth', 'R']]
data_1 = data_1.drop_duplicates()


data_2 = pd.read_csv(col_uv_f + '/final.modified_unmodified_new', sep='\t', header=None)
data_2.columns = ['transcript_id', 'Nucleotide', 'Modified_mutations', 'Modified_effective_depth',
                     'Untreated_mutations', 'Untreated_effective_depth', 'R']
data_2 = data_2[['transcript_id', 'Nucleotide', 'Modified_mutations', 'Modified_effective_depth',
                       'Untreated_mutations', 'Untreated_effective_depth', 'R']]
data_2 = data_2.drop_duplicates()

data_3 = pd.read_csv(uvr_uv_z + '/final.modified_unmodified_new', sep='\t', header=None)
data_3.columns = ['transcript_id', 'Nucleotide', 'Modified_mutations', 'Modified_effective_depth',
                     'Untreated_mutations', 'Untreated_effective_depth', 'R']
data_3 = data_3[['transcript_id', 'Nucleotide', 'Modified_mutations', 'Modified_effective_depth',
                       'Untreated_mutations', 'Untreated_effective_depth', 'R']]
data_3 = data_3.drop_duplicates()


data_4 = pd.read_csv(uvr_uv_f + '/final.modified_unmodified_new', sep='\t', header=None)
data_4.columns = ['transcript_id', 'Nucleotide', 'Modified_mutations', 'Modified_effective_depth',
                     'Untreated_mutations', 'Untreated_effective_depth', 'R']
data_4 = data_4[['transcript_id', 'Nucleotide', 'Modified_mutations', 'Modified_effective_depth',
                       'Untreated_mutations', 'Untreated_effective_depth', 'R']]
data_4 = data_4.drop_duplicates()

path = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/result/col/'
savepath = '/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT/'
shape_data_coil = pd.read_csv(path+'/summary_result_merge_50_1_new.csv',sep='\t')
shape_data_coil = shape_data_coil.loc[(shape_data_coil['hit_f']>2)&(shape_data_coil['hit_z']>2),:]
shape_data_coil.index=range(len(shape_data_coil))
shape_data_coil['up_down']= shape_data_coil['delta'].map(lambda x: 'up' if x>0 else 'down')
shape_data_coil = shape_data_coil.reset_index()
trasnscript_list=list(shape_data_coil['transcript_id'])
shape_data_coil[['index','transcript_id','up_down','start']].to_csv(savepath+'index.csv',sep='\t',index=False)
for i in range(len(trasnscript_list)):
    print(i)
    trasnscript=trasnscript_list[i]
    f = open(savepath+'/fasta/'+str(i)+'.fasta','w')
    start = int(shape_data_coil.loc[i, 'start'])
    end = int(shape_data_coil.loc[i, 'end'])
    location = shape_data_coil.loc[i, 'location_start']
    up_down =shape_data_coil.loc[i, 'up_down']
    seq = shape_data_coil.loc[i, 'nucleotide'].replace(',','')
    f.write('>'+trasnscript+'_'+location+'_'+up_down+'_'+str(start)+'_'+str(end)+'\n')
    f.write(seq+'\n')
    f.close()
    data_z_ = list(data_1.loc[data_1['transcript_id'] == trasnscript, 'R'])[0].split(',')[start:end]
    data_f_ = list(data_2.loc[data_2['transcript_id'] == trasnscript, 'R'])[0].split(',')[start:end]
    #data_z_ = list(data_3.loc[data_3['transcript_id'] == trasnscript, 'R'])[0].split(',')[start:end]
    #data_f_ = list(data_4.loc[data_4['transcript_id'] == trasnscript, 'R'])[0].split(',')[start:end]
    data_z = pd.DataFrame(columns=['Positions', 'Reactivity'])
    data_z['Positions'] = [j for j in range(len(data_z_))]
    data_z['Reactivity'] = data_z_
    data_z = data_z[['Positions', 'Reactivity']]
    data_f = pd.DataFrame(columns=['Positions', 'Reactivity'])
    data_f['Positions'] = [j for j in range(len(data_f_))]
    data_f['Reactivity'] = data_f_
    data_f = data_f[['Positions', 'Reactivity']]
    data_z.to_csv(savepath+'/shape/'+str(i)+'_z.shape',sep='\t',index=False,header=None)
    data_f.to_csv(savepath+'/shape/'+str(i)+'_f.shape',sep='\t',index=False,header=None)
```
这里以WT样本为例，输出文件夹为/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT/fasta和/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT/shape

我们可以path，savepath等参数处理UVR8样本。

#### 4b) 预测RNA structure

我们利用RNAstructure中ShapeKnots和ct2dot函数，基于结构改变区域的序列和reactivity文件进行二级结构预测。

```sh
#!/bin/sh
fasta=$"/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT/fasta/"
shape=$"/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT/shape/"
ct_path=$"/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT/dot/"
source /BioII/lulab_b/containers/singularity/wrappers/bashrc
for i in `seq 0 1580`
do
echo $i
ShapeKnots $fasta/"$i".fasta $ct_path/"$i"_z.ct -sh $shape/"$i"_z.shape -sm 1.9 -si -0.7
ShapeKnots $fasta/"$i".fasta $ct_path/"$i"_f.ct -sh $shape/"$i"_f.shape -sm 1.9 -si -0.7
ct2dot $ct_path/"$i"_z.ct 1 $ct_path/"$i"_z.dt
ct2dot $ct_path/"$i"_f.ct 1 $ct_path/"$i"_f.dt
done

fasta=$"/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/UVR/fasta/"
shape=$"/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/UVR/shape/"
ct_path=$"/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/UVR/dot/"
source /BioII/lulab_b/containers/singularity/wrappers/bashrc
for i in `seq 0 1313`
do
echo $i
ShapeKnots $fasta/"$i".fasta $ct_path/"$i"_z.ct -sh $shape/"$i"_z.shape -sm 1.9 -si -0.7
ShapeKnots $fasta/"$i".fasta $ct_path/"$i"_f.ct -sh $shape/"$i"_f.shape -sm 1.9 -si -0.7
ct2dot $ct_path/"$i"_z.ct 1 $ct_path/"$i"_z.dt
ct2dot $ct_path/"$i"_f.ct 1 $ct_path/"$i"_f.dt
done
```

我们可以获得每个结构变化区域在UV-和UV+下的RNA二级结构，输出文件夹为/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT/dot。其中*\_f.ct为UV-下的RNA结构，*\_z.ct为UV+下的RNA结构。

#### 4c) 数据汇总

我们对UV-下上调的结构变化区域和下调结构变化区域进行motif富集分析，需要对UV-下上调/下调的所有结构变化区域进行汇总。

脚本参考/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/3.merge.py

```python
import pandas as pd
path='/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT/'
index = pd.read_csv(path+'index.csv',sep='\t')
index_up = list(index.loc[index['up_down']=='up','index'])
index_down = list(index.loc[index['up_down']=='down','index'])

f_new = open(path + 'merge_up_nouv.fa', 'w')
for i in range(len(index_up)):
    try:
        f = open(path + '/dot/' + str(index_up[i]) + '_f.dt')
        j=0
        for line in f:
            # print(line)
            if j==0:
                transcript = line.split(' ')[-1]
                f_new.write('>'+transcript)
            else:
                f_new.write(line)
            j=j+1
        f.close()
    except FileNotFoundError:
        continue
f_new.close()

f_new = open(path + 'merge_down_nouv.fa', 'w')
for i in range(len(index_down)):
    try:
        f = open(path + '/dot/' + str(index_down[i]) + '_f.dt')
        j=0
        for line in f:
            # print(line)
            if j==0:
                transcript = line.split(' ')[-1]
                f_new.write('>'+transcript)
            else:
                f_new.write(line)
            j=j+1
        f.close()
    except FileNotFoundError:
        continue
f_new.close()

path='/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/UVR/'
index = pd.read_csv(path+'index.csv',sep='\t')
index_up = list(index.loc[index['up_down']=='up','index'])
index_down = list(index.loc[index['up_down']=='down','index'])

f_new = open(path + 'merge_up_nouv.fa', 'w')
for i in range(len(index_up)):
    try:
        f = open(path + '/dot/' + str(index_up[i]) + '_f.dt')
        j=0
        for line in f:
            # print(line)
            if j==0:
                transcript = line.split(' ')[-1]
                f_new.write('>'+transcript)
            else:
                f_new.write(line)
            j=j+1
        f.close()
    except FileNotFoundError:
        continue
f_new.close()

f_new = open(path + 'merge_down_nouv.fa', 'w')
for i in range(len(index_down)):
    try:
        f = open(path + '/dot/' + str(index_down[i]) + '_f.dt')
        j=0
        for line in f:
            # print(line)
            if j==0:
                transcript = line.split(' ')[-1]
                f_new.write('>'+transcript)
            else:
                f_new.write(line)
            j=j+1
        f.close()
    except FileNotFoundError:
        continue
f_new.close()
```

WT输出文件为/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT/merge_up_nouv.fa，/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT/merge_down_uv.fa。

#### 4d) motif discovery

利用BEAM，对motif进行富集分析。

```sh
cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT
source /BioII/lulab_b/containers/singularity/wrappers/bashrc

java -jar /BioII/lulab_b/liuxiaofan/software/BEAR_encoder/jar/BEAR_encoder.jar -i merge_down_nouv.fa -o merge_down_nouv_ready.fa
java -jar /BioII/lulab_b/liuxiaofan/software/BEAR_encoder/jar/BEAR_encoder.jar -i merge_up_nouv.fa -o merge_up_nouv_ready.fa

java -jar /BioII/lulab_b/liuxiaofan/software/beam-master/BEAM_release2.5.0.jar -f merge_down_nouv_ready.fa -w 10 -W 50 -M 100
java -jar /BioII/lulab_b/liuxiaofan/software/beam-master/BEAM_release2.5.0.jar -f merge_up_nouv_ready.fa -w 10 -W 50 -M 100

cd /BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/UVR
source /BioII/lulab_b/containers/singularity/wrappers/bashrc

java -jar /BioII/lulab_b/liuxiaofan/software/BEAR_encoder/jar/BEAR_encoder.jar -i merge_down_nouv.fa -o merge_down_nouv_ready.fa
java -jar /BioII/lulab_b/liuxiaofan/software/BEAR_encoder/jar/BEAR_encoder.jar -i merge_up_nouv.fa -o merge_up_nouv_ready.fa

java -jar /BioII/lulab_b/liuxiaofan/software/beam-master/BEAM_release2.5.0.jar -f merge_down_nouv_ready.fa -w 10 -W 50 -M 100
java -jar /BioII/lulab_b/liuxiaofan/software/beam-master/BEAM_release2.5.0.jar -f merge_up_nouv_ready.fa -w 10 -W 50 -M 100
```

得到输出文件夹/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT/results和/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/UVR/results


#### 4e) motif汇总

我们根据计算得到f结构moti进行过滤、汇总和绘图。

**1.过滤结构motif**

我们以WT样本上调结构变化区域找的结构motif为例，说明过滤规则。我们在上面步骤计算得到结果文件夹/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT/results/merge_up_nouv_ready。merge_up_nouv_ready_summary.txt文件为结构motif汇总。

我们过滤规则如下：

* motif宽度大于5
* pvalueMW:pvalue of MannWhitney U Test against,the null hypothesis that the two samples of partial scores of the RNA with the motif and the partials scores of top scoring background with the model PFM，要求pvalue<0.05
* AUC:derived from the U statistic, it rates the classifying power of the motif，要求AUC>0.65
* coverage:fraction of RNA in the input containing the motif,要求coverage>20%

**2.结构motif汇总**

我们根据merge_up_nouv_ready_summary.txt文件可以得到过滤后motif的pvalueMW，AUC，qBEAR。

另外我们在/BioII/lulab_b/liuxiaofan/project/ribo-shape/result_liulab_batch4_new/07.BEAM/WT/results/merge_up_nouv_ready/benchmark/motifs/\*\_ready\_m1\_run1.txt中找到每个motif对应的结果改变区域的具体信息。

![7.png](/png/7.png)

**3.绘图**

我们利用**weblogo**进行结构motif绘图。我们对每个过滤后的motif进行绘图。

```sh
weblogo -a 'ZAQXSWCDEVFRBGTNHY' -f merge_up_ck_ready_m1_run1_wl.fa  -D fasta \
-o m1.pdf -F pdf --composition="none" \
-C red ZAQ 'Stem' -C blue XSW 'Loop' -C forestgreen CDE 'InternalLoop' \
-C orange VFR 'StemBranch' -C DarkOrange B 'Bulge' \
-C lime G 'BulgeBranch' -C purple T 'Branching' \
-C limegreen NHY 'InternalLoopBranch'
```
![8.png](/png/8.png)

图片中为每个位点qBEAR，不同字母的大小代表这个位点为该字母的可能性（因为motif允许一定错配），字母含义为motif结构编码，Z:big Stem,A:medial Stem,Q:small Stem,X:big Loop,S:medial Loop,W:small Loop,C:big InternalLoop,D:medial InternalLoop,E:small InternalLoop,V:big StemBranch,F:medial StemBranch,R:small StemBranch,B:Bulge,G:BulgeBranch,T:Branching,N:big InternalLoopBranch,H:medial InternalLoopBranch,Y:small InternalLoopBranch
